File: canonforge-omega/omega_runtime/successor_evidence_r186.py
from __future__ import annotations

import hashlib
import json
from pathlib import Path
from typing import Any, Dict, Iterable, Mapping

from .successor_gate import REQUIRED_CONNECTIONS_R183, SUCCESSOR_METRIC_WEIGHTS_R183
from .warp_candidate import canonical_sha256, validate_candidate_capsule


EVIDENCE_REVISION_R186 = "R186"
EVIDENCE_JOB_SCHEMA_R186 = "OMEGA_SUCCESSOR_EVIDENCE_JOB_R186"
EVIDENCE_RECEIPT_SCHEMA_R186 = "OMEGA_EXECUTION_DERIVED_SUCCESSOR_EVIDENCE_R186"


CAPABILITY_FILES = {
    "R170_REFERENCE_COMPUTE": "cloudflare/omega-v6-worker/src/compute/computeTruthR170.ts",
    "R172_HETEROGENEOUS_VALIDATION": "cloudflare/omega-v6-worker/src/validation/validationFabricR172.ts",
    "R173_CROSS_RUNTIME_PARITY": "cloudflare/omega-v6-worker/src/validation/crossRuntimeParityR173.ts",
    "R174_FEDERATED_ORGANS": "cloudflare/omega-v6-worker/src/federation/federatedOrganFabricR174.ts",
    "R175_INDEPENDENT_SOLVER": "cloudflare/omega-v6-worker/src/validation/independentSolverR175.ts",
    "R176_WARP_COMPUTE": "cloudflare/omega-v6-worker/src/swarm/warpComputationR176.ts",
    "R178_BUILD_CANDIDATE": "cloudflare/omega-v6-worker/src/swarm/warpBuildCandidateR178.ts",
    "R179_AI_SAI": "cloudflare/omega-v6-worker/src/intelligence/saiAiFusionR179.ts",
    "R180_LOAD_GOVERNOR": "cloudflare/omega-v6-worker/src/swarm/swarmGovernorR180.ts",
    "R181_LIVE_ACCEPTANCE": "cloudflare/omega-v6-worker/src/acceptance/liveAcceptanceR181.ts",
    "R183_SUCCESSOR_GATE": "cloudflare/omega-v6-worker/src/swarm/successorGateR183.ts",
    "R184_IMPROVEMENT_DISCOVERY": "cloudflare/omega-v6-worker/src/swarm/improvementDiscoveryR184.ts",
    "R185_172_CLOUD_FEDERATION": "cloudflare/omega-v6-worker/src/swarm/cloudSwarmR185.ts",
    "SOVEREIGN_SELF_BUILD": "omega_runtime/self_build.py",
    "SOVEREIGN_AGENT": "scripts/omega_sovereign_agent.py",
    "SAI_B059": "omega_runtime/sai_b059.py",
}


def _sha(value: Any) -> str:
    text = value if isinstance(value, str) else json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False, default=str)
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def _exit_ok(value: Mapping[str, Any] | None) -> bool:
    return isinstance(value, Mapping) and int(value.get("exit_code", 1)) == 0


def _elapsed(value: Mapping[str, Any] | None) -> float:
    try:
        return max(0.0, float((value or {}).get("elapsed_seconds", 0.0)))
    except (TypeError, ValueError):
        return 0.0


def capabilities_from_root(root: Path) -> list[str]:
    return sorted(name for name, rel in CAPABILITY_FILES.items() if (root / rel).exists())


def stage_receipts(stage_evidence: Mapping[str, Any]) -> tuple[list[Dict[str, Any]], Dict[str, str]]:
    stages: list[Dict[str, Any]] = []
    hashes: Dict[str, str] = {}
    for kind in ("run_tests", "build_vite", "wrangler_dry_run", "verify_candidate"):
        evidence = stage_evidence.get(kind)
        if not isinstance(evidence, Mapping):
            stages.append({"kind": kind, "state": "MISSING", "evidenceSha256": None})
            continue
        digest = canonical_sha256(dict(evidence))
        hashes[kind] = digest
        stages.append({"kind": kind, "state": "VERIFIED", "evidenceSha256": digest})
    return stages, hashes


def execution_metric_vector(
    *,
    test_result: Mapping[str, Any] | None,
    typecheck_result: Mapping[str, Any] | None,
    dry_run_result: Mapping[str, Any] | None,
    verify_result: Mapping[str, Any] | None,
    capabilities: Iterable[str],
    changed_files: Iterable[str],
    objective: str,
    connections: Mapping[str, bool],
    clean_workspace: bool,
) -> Dict[str, float]:
    tests_ok = _exit_ok(test_result)
    typecheck_ok = _exit_ok(typecheck_result)
    dry_ok = _exit_ok(dry_run_result)
    verify_ok = bool(verify_result)
    caps = list(dict.fromkeys(str(x) for x in capabilities))
    changes = [str(x) for x in changed_files if str(x)]
    proof_passes = sum([tests_ok, typecheck_ok, dry_ok, verify_ok])
    federation = sum(bool(connections.get(k)) for k in ("canonicalEdge", "swarm", "genesisMachine", "opticalMachine")) / 4.0
    total_elapsed = _elapsed(test_result) + _elapsed(typecheck_result) + _elapsed(dry_run_result)
    # Deterministic resource/interaction references; same law is used for predecessor and successor.
    interactive_eff = 1.0 / (1.0 + total_elapsed / 120.0)
    resource_eff = 1.0 / (1.0 + total_elapsed / 300.0)
    return {
        "correctness": 1.0 if tests_ok else 0.0,
        "regressionSafety": sum([tests_ok, typecheck_ok, dry_ok]) / 3.0,
        "missionFit": 1.0 if objective.strip() and changes else 0.0,
        "proofCoverage": proof_passes / 4.0,
        "capabilityCoverage": min(1.0, len(caps) / max(1, len(CAPABILITY_FILES))),
        "interactiveEfficiency": interactive_eff,
        "resourceEfficiency": resource_eff,
        "recoverability": 1.0 if clean_workspace and tests_ok else 0.0,
        "continuity": 1.0 if "R180_LOAD_GOVERNOR" in caps and "R183_SUCCESSOR_GATE" in caps and tests_ok else 0.0,
        "federationConnectivity": federation,
        "sovereignConnectivity": 1.0 if connections.get("sovereignHost") else 0.0,
    }


def predecessor_metric_vector(
    *,
    benchmark: Mapping[str, Any],
    capabilities: Iterable[str],
    connections: Mapping[str, bool],
) -> Dict[str, float]:
    return execution_metric_vector(
        test_result=benchmark.get("tests") if isinstance(benchmark.get("tests"), Mapping) else None,
        typecheck_result=benchmark.get("typecheck") if isinstance(benchmark.get("typecheck"), Mapping) else None,
        dry_run_result=benchmark.get("dryRun") if isinstance(benchmark.get("dryRun"), Mapping) else None,
        verify_result={"predecessor": True} if benchmark.get("verified") is True else None,
        capabilities=capabilities,
        changed_files=[],
        objective="",
        connections=connections,
        clean_workspace=benchmark.get("cleanWorkspace") is True,
    )


def build_execution_evidence_r186(
    *,
    candidate: Mapping[str, Any],
    predecessor_sha: str,
    predecessor_metrics: Mapping[str, Any],
    predecessor_capabilities: Iterable[str],
    successor_sha: str,
    successor_branch: str,
    successor_metrics: Mapping[str, Any],
    successor_capabilities: Iterable[str],
    stage_evidence: Mapping[str, Any],
    connections: Mapping[str, bool],
    connection_receipts: Mapping[str, Any],
    changed_files: Iterable[str],
    diff_text: str,
) -> Dict[str, Any]:
    valid = validate_candidate_capsule(dict(candidate))
    stages, stage_hashes = stage_receipts(stage_evidence)
    changed = list(dict.fromkeys(str(x) for x in changed_files if str(x)))
    diff_sha = _sha(diff_text)
    artifact_core = {
        "sourceCommitSha": successor_sha,
        "candidateSha256": valid["capsuleSha256"],
        "changedFiles": changed,
        "diffSha256": diff_sha,
        "stageEvidence": stage_hashes,
    }
    artifact_sha = canonical_sha256(artifact_core)
    metric_evidence = {
        key: canonical_sha256({
            "metric": key,
            "predecessor": predecessor_metrics.get(key),
            "successor": successor_metrics.get(key),
            "basis": "R186_EXECUTION_DERIVED",
        })
        for key in SUCCESSOR_METRIC_WEIGHTS_R183
    }
    connection_evidence = {
        key: canonical_sha256({
            "connection": key,
            "observed": connections.get(key) is True,
            "receipt": connection_receipts.get(key),
        })
        for key in REQUIRED_CONNECTIONS_R183
        if connections.get(key) is True
    }
    mission = {
        "schema": "OMEGA_EXACT_BUILD_MISSION_R186",
        "revision": EVIDENCE_REVISION_R186,
        "exactBuildDevelopment": True,
        "objective": str(valid.get("objective") or ""),
        "candidateSha256": valid["capsuleSha256"],
        "predecessorCanonicalGitSha": predecessor_sha,
    }
    mission_sha = canonical_sha256(mission)
    design_output = {
        "schema": "OMEGA_IMPROVEMENT_DESIGN_OUTPUT_R184",
        "sourceBranch": successor_branch,
        "sourceCommitSha": successor_sha,
        "predecessorCanonicalGitSha": predecessor_sha,
        "candidateSha256": valid["capsuleSha256"],
        "missionSha256": mission_sha,
        "diffSha256": diff_sha,
        "artifactSha256": artifact_sha,
        "changedFiles": changed,
        "evidenceSha256": list(stage_hashes.values()),
        "metricEvidence": metric_evidence,
        "connectionEvidence": connection_evidence,
        "canonicalMutation": False,
        "githubMutationAuthorized": False,
        "deploymentAuthorized": False,
        "promotionAuthorized": False,
    }
    r183_payload = {
        "candidate": valid,
        "predecessor": {
            "canonicalGitSha": predecessor_sha,
            "metrics": dict(predecessor_metrics),
            "capabilities": list(predecessor_capabilities),
        },
        "successor": {
            "metrics": dict(successor_metrics),
            "capabilities": list(successor_capabilities),
            "verification": {"stages": stages},
            "connections": {key: connections.get(key) is True for key in REQUIRED_CONNECTIONS_R183},
        },
        "mission": mission,
        "designOutput": design_output,
    }
    core = {
        "schema": EVIDENCE_RECEIPT_SCHEMA_R186,
        "revision": EVIDENCE_REVISION_R186,
        "candidateId": valid["candidateId"],
        "candidateSha256": valid["capsuleSha256"],
        "predecessorCanonicalGitSha": predecessor_sha,
        "successorGitSha": successor_sha,
        "successorBranch": successor_branch,
        "predecessorMetrics": dict(predecessor_metrics),
        "successorMetrics": dict(successor_metrics),
        "predecessorCapabilities": list(predecessor_capabilities),
        "successorCapabilities": list(successor_capabilities),
        "verificationStages": stages,
        "connections": {key: connections.get(key) is True for key in REQUIRED_CONNECTIONS_R183},
        "changedFiles": changed,
        "diffSha256": diff_sha,
        "artifactSha256": artifact_sha,
        "metricEvidence": metric_evidence,
        "connectionEvidence": connection_evidence,
        "r183R184Payload": r183_payload,
        "authority": "AUTHENTICATED_EXECUTION_EVIDENCE_NOT_CANON",
        "canonicalMutation": False,
        "promotionAuthorized": False,
    }
    return {**core, "receiptSha256": canonical_sha256(core)}

File: canonforge-omega/cloudflare/omega-v6-worker/src/swarm/successorEvidenceR186.ts
import { AnyObj, SwarmEnv, jsonResponse, sha } from "./swarmCoreR169";
import { canonicalShaR178 } from "./warpBuildCandidateR178";
import { evaluateSuccessorR183, SUCCESSOR_METRIC_WEIGHTS_R183 } from "./successorGateR183";
import { discoverImprovementR184 } from "./improvementDiscoveryR184";

export const SUCCESSOR_EVIDENCE_REVISION_R186 = "R186";
export const SUCCESSOR_EVIDENCE_SCHEMA_R186 = "OMEGA_EXECUTION_DERIVED_SUCCESSOR_EVIDENCE_R186";

const REQUIRED_CONNECTIONS = ["canonicalEdge", "swarm", "genesisMachine", "opticalMachine", "sovereignHost", "sai"] as const;
const VERIFY_STAGES = ["run_tests", "build_vite", "wrangler_dry_run", "verify_candidate"] as const;

type CanonicalFetch = (request: Request, env: any, ctx: any) => Promise<Response>;

const clamp01 = (v: number) => Math.max(0, Math.min(1, Number.isFinite(v) ? v : 0));
const arr = (v: any): string[] => Array.isArray(v) ? [...new Set(v.map((x) => String(x || "").trim()).filter(Boolean))] : [];
const exitOk = (v: any): boolean => Boolean(v && (Number(v.exit_code) === 0 || v.passed === true || v.ok === true || String(v.state || "").toUpperCase() === "VERIFIED"));
const elapsed = (v: any): number => Math.max(0, Number(v?.elapsed_seconds || v?.elapsedSeconds || 0) || 0);

async function callCanonical(request: Request, env: any, ctx: any, canonicalFetch: CanonicalFetch, path: string): Promise<AnyObj> {
  try {
    const u = new URL(request.url); u.pathname = path; u.search = "";
    const response = await canonicalFetch(new Request(u.toString(), { method: "GET", headers: { accept: "application/json" } }), env, ctx);
    const body = await response.json().catch(() => null) as AnyObj | null;
    return { reachable: response.ok, status: response.status, body };
  } catch (error) {
    return { reachable: false, status: 0, error: error instanceof Error ? error.message : String(error), body: null };
  }
}

async function callService(binding: any, service: string): Promise<AnyObj> {
  if (!binding?.fetch) return { reachable: false, service, code: "SERVICE_BINDING_UNAVAILABLE" };
  try {
    const response = await binding.fetch(new Request(`https://${service}.internal/api/health`, { method: "GET", headers: { accept: "application/json" } }));
    const body = await response.json().catch(() => null) as AnyObj | null;
    return { reachable: response.ok && body?.ok === true, status: response.status, service, body };
  } catch (error) {
    return { reachable: false, status: 0, service, error: error instanceof Error ? error.message : String(error) };
  }
}

async function swarmProbe(env: SwarmEnv): Promise<AnyObj> {
  if (!env.OMEGA_SWARM_CELL) return { reachable: false, code: "OMEGA_SWARM_CELL_BINDING_UNAVAILABLE" };
  try {
    const binding = env.OMEGA_SWARM_CELL;
    const stub = binding.get(binding.idFromName("omega-cloud-r185-001"));
    const response = await stub.fetch(new Request("https://r186.internal/state", { method: "GET" }));
    const body = await response.json().catch(() => null) as AnyObj | null;
    return { reachable: response.ok, status: response.status, body };
  } catch (error) {
    return { reachable: false, status: 0, error: error instanceof Error ? error.message : String(error) };
  }
}

export async function collectLiveConnectionsR186(request: Request, env: any, ctx: any, canonicalFetch: CanonicalFetch): Promise<AnyObj> {
  const [hybrid, saiStatus, genesis, optical, swarm] = await Promise.all([
    callCanonical(request, env, ctx, canonicalFetch, "/api/hybrid/status"),
    callCanonical(request, env, ctx, canonicalFetch, "/api/sai/status"),
    callService(env.OMEGA_GENESIS_MACHINE, "omega-genesis-machine-r115"),
    callService(env.OMEGA_OPTICAL_MACHINE, "omega-optical-machine-r115"),
    swarmProbe(env),
  ]);
  const hb = hybrid.body || {};
  const sai = saiStatus.body || {};
  const canonicalGitSha = String(env?.CANONICAL_GIT_SHA || "").trim();
  const observed = {
    canonicalEdge: /^[a-f0-9]{40,64}$/i.test(canonicalGitSha),
    swarm: swarm.reachable === true,
    genesisMachine: genesis.reachable === true,
    opticalMachine: optical.reachable === true,
    sovereignHost: Boolean(hb.pcOnline ?? hb.pc_online) && Boolean(hb.authenticated ?? hb.proof?.authenticated ?? hb.authenticated_heartbeat),
    sai: Boolean(sai.passed === true || sai.state === "B059_PRESENT_VERIFICATION_REQUIRED" || sai.release === "OMEGA SAI B059"),
  };
  const receipts: AnyObj = { canonicalEdge: { canonicalGitSha }, swarm, genesisMachine: genesis, opticalMachine: optical, sovereignHost: hybrid, sai: saiStatus };
  const evidence: AnyObj = {};
  for (const key of REQUIRED_CONNECTIONS) if ((observed as AnyObj)[key] === true) evidence[key] = await canonicalShaR178({ key, observed: true, receipt: receipts[key] });
  return { observed, receipts, evidence, complete: REQUIRED_CONNECTIONS.every((key) => (observed as AnyObj)[key] === true) };
}

function executionMetrics(raw: AnyObj, capabilities: string[], changedFiles: string[], objective: string, connections: AnyObj): AnyObj {
  const tests = raw.run_tests || raw.tests || {};
  const typecheck = raw.build_vite || raw.typecheck || {};
  const dry = raw.wrangler_dry_run || raw.dryRun || {};
  const verify = raw.verify_candidate || raw.verifyCandidate || {};
  const testsOk = exitOk(tests), typeOk = exitOk(typecheck), dryOk = exitOk(dry), verifyOk = exitOk(verify);
  const proofPasses = [testsOk, typeOk, dryOk, verifyOk].filter(Boolean).length;
  const totalElapsed = elapsed(tests) + elapsed(typecheck) + elapsed(dry);
  const federation = ["canonicalEdge", "swarm", "genesisMachine", "opticalMachine"].filter((k) => connections[k] === true).length / 4;
  const continuityPresent = capabilities.some((x) => /R182|CONTINUITY/i.test(x)) && capabilities.some((x) => /R183|SUCCESSOR/i.test(x));
  const maxCapabilities = Math.max(1, Number(raw.capabilityScale || 32));
  return {
    correctness: testsOk ? 1 : 0,
    regressionSafety: [testsOk, typeOk, dryOk].filter(Boolean).length / 3,
    missionFit: objective.trim() && changedFiles.length ? 1 : 0,
    proofCoverage: proofPasses / 4,
    capabilityCoverage: clamp01(capabilities.length / maxCapabilities),
    interactiveEfficiency: 1 / (1 + totalElapsed / 120),
    resourceEfficiency: 1 / (1 + totalElapsed / 300),
    recoverability: raw.cleanWorkspace === true && testsOk ? 1 : 0,
    continuity: continuityPresent && testsOk ? 1 : 0,
    federationConnectivity: federation,
    sovereignConnectivity: connections.sovereignHost === true ? 1 : 0,
  };
}

async function stageTruth(raw: AnyObj): Promise<AnyObj> {
  const map: AnyObj = {
    run_tests: raw.run_tests || raw.tests || {},
    build_vite: raw.build_vite || raw.typecheck || {},
    wrangler_dry_run: raw.wrangler_dry_run || raw.dryRun || {},
    verify_candidate: raw.verify_candidate || raw.verifyCandidate || {},
  };
  const stages: AnyObj[] = [];
  for (const kind of VERIFY_STAGES) {
    const evidence = map[kind];
    stages.push({ kind, state: exitOk(evidence) ? "VERIFIED" : "FAILED", evidenceSha256: await canonicalShaR178(evidence) });
  }
  return { stages, complete: stages.every((x) => x.state === "VERIFIED") };
}

export async function collectSuccessorEvidenceR186(body: AnyObj, request: Request, env: any, ctx: any, canonicalFetch: CanonicalFetch): Promise<AnyObj> {
  const candidate = body.candidate || {};
  const predecessor = body.predecessor || {};
  const source = body.source || {};
  const predecessorExecution = body.predecessorExecution || {};
  const successorExecution = body.successorExecution || {};
  const objective = String(candidate.objective || body.objective || "").trim();
  const changedFiles = arr(source.changedFiles);
  const predecessorCaps = arr(predecessor.capabilities);
  const successorCaps = arr(body.successorCapabilities);
  const live = await collectLiveConnectionsR186(request, env, ctx, canonicalFetch);
  const predecessorMetrics = executionMetrics(predecessorExecution, predecessorCaps, [], "", live.observed);
  const successorMetrics = executionMetrics(successorExecution, successorCaps, changedFiles, objective, live.observed);
  const verification = await stageTruth(successorExecution);
  const metricEvidence: AnyObj = {};
  for (const key of Object.keys(SUCCESSOR_METRIC_WEIGHTS_R183)) metricEvidence[key] = await canonicalShaR178({ metric: key, predecessor: predecessorMetrics[key], successor: successorMetrics[key], basis: "R186_EXECUTION_DERIVED" });

  const mission = {
    schema: "OMEGA_EXACT_BUILD_MISSION_R186",
    revision: SUCCESSOR_EVIDENCE_REVISION_R186,
    exactBuildDevelopment: true,
    objective,
    candidateSha256: candidate.capsuleSha256 || null,
    predecessorCanonicalGitSha: predecessor.canonicalGitSha || null,
  };
  const missionSha256 = await canonicalShaR178(mission);
  const diffText = String(source.diff || "");
  const diffSha256 = await sha(diffText);
  const stageHashes = verification.stages.map((x: AnyObj) => x.evidenceSha256);
  const artifactSha256 = await canonicalShaR178({ sourceCommitSha: source.commitSha, candidateSha256: candidate.capsuleSha256, changedFiles, diffSha256, stageHashes });
  const designOutput = {
    schema: "OMEGA_IMPROVEMENT_DESIGN_OUTPUT_R184",
    sourceBranch: source.branch || null,
    sourceCommitSha: source.commitSha || null,
    predecessorCanonicalGitSha: predecessor.canonicalGitSha || null,
    candidateSha256: candidate.capsuleSha256 || null,
    missionSha256,
    diffSha256,
    artifactSha256,
    changedFiles,
    evidenceSha256: stageHashes,
    metricEvidence,
    connectionEvidence: live.evidence,
    canonicalMutation: false,
    githubMutationAuthorized: false,
    deploymentAuthorized: false,
    promotionAuthorized: false,
  };
  const evaluationPayload = {
    candidate,
    predecessor: { canonicalGitSha: predecessor.canonicalGitSha, metrics: predecessorMetrics, capabilities: predecessorCaps },
    successor: { metrics: successorMetrics, capabilities: successorCaps, verification: { stages: verification.stages }, connections: live.observed },
    mission,
    designOutput,
    policy: body.policy || {},
  };
  const successorEvaluation = await evaluateSuccessorR183(evaluationPayload);
  const improvementDiscovery = await discoverImprovementR184(evaluationPayload);
  const core = {
    schema: SUCCESSOR_EVIDENCE_SCHEMA_R186,
    revision: SUCCESSOR_EVIDENCE_REVISION_R186,
    state: improvementDiscovery.discoveredSuperiorDesign === true ? "SUPERIOR_DESIGN_DISCOVERED" : "CONTINUE_SELF_DEVELOPMENT",
    executionDerivedMetrics: true,
    liveConnectionCollection: true,
    predecessorMetrics,
    successorMetrics,
    verification,
    liveConnections: live,
    metricEvidence,
    designOutput,
    successorEvaluation,
    improvementDiscovery,
    continuation: improvementDiscovery.continuation || successorEvaluation.continuation || null,
    canonicalMutation: false,
    promotionAuthorized: false,
    authority: "R186_EXECUTION_DERIVED_EVIDENCE_NOT_RELEASE_AUTHORITY",
    boundary: "R186 derives scores from raw execution receipts and current live connection probes. Raw predecessor/successor execution receipts and source identity are inputs; callers cannot directly supply R183 metric scores. Release remains proof-gated.",
  };
  return { ok: true, ...core, receiptSha256: await canonicalShaR178(core) };
}

export async function handleSuccessorEvidenceR186(request: Request, env: any, ctx: any, canonicalFetch: CanonicalFetch): Promise<Response> {
  const path = new URL(request.url).pathname;
  if (request.method === "GET" && path === "/api/swarm/evidence/r186/manifest") return jsonResponse({
    ok: true,
    schema: "OMEGA_SUCCESSOR_EVIDENCE_MANIFEST_R186",
    revision: SUCCESSOR_EVIDENCE_REVISION_R186,
    derivesMetricsFromRawExecutionReceipts: true,
    collectsLiveConnections: REQUIRED_CONNECTIONS,
    feedsSuccessorGate: "R183",
    feedsImprovementDiscovery: "R184",
    continuationOnIncompleteOrInferior: true,
    automaticProductionPromotion: false,
    canonicalMutation: false,
  });
  if (request.method === "POST" && path === "/api/swarm/evidence/r186/collect") {
    const body = await request.json().catch(() => ({})) as AnyObj;
    try { return jsonResponse(await collectSuccessorEvidenceR186(body, request, env, ctx, canonicalFetch)); }
    catch (error) { return jsonResponse({ ok: false, schema: SUCCESSOR_EVIDENCE_SCHEMA_R186, revision: SUCCESSOR_EVIDENCE_REVISION_R186, error: error instanceof Error ? error.message : String(error), canonicalMutation: false, promotionAuthorized: false }, 400); }
  }
  return jsonResponse({ ok: false, code: "R186_EVIDENCE_ROUTE_NOT_FOUND", canonicalMutation: false }, 404);
}
